packages/core/src/browser/dialogs.ts

```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
     return safeToExit === true;
 }
 
+export class ConfirmDialogSaveProps extends ConfirmDialogProps {
+    readonly save: string;
+    performSave: () => void;
+}
+
+export class ConfirmDialogSave extends ConfirmDialog {
+
+    protected saveButton: HTMLButtonElement | undefined;
+    constructor(
+        @inject(ConfirmDialogSaveProps) protected override readonly props: ConfirmDialogSaveProps
+    ) {
+        super(props);
+        this.contentNode.appendChild(this.createMessageNode(this.props.msg));
+        // reorder buttons
+        this.controlPanel.childNodes.forEach(child => this.controlPanel.removeChild(child));
+        [this.acceptButton, this.closeButton].forEach(child => {
+            if (typeof child !== 'undefined') {
+                this.controlPanel.appendChild(child);
+            }
+        });
+        this.appendSaveButton(props.save).addEventListener('click', () => { props.performSave(); this.acceptButton?.click(); });
+    }
+
+    protected appendSaveButton(text: string = Dialog.OK): HTMLButtonElement {
+        this.saveButton = this.createButton(text);
+        this.controlPanel.appendChild(this.saveButton);
+        this.saveButton.classList.add('main');
+        return this.saveButton;
+    }
+
+}
+
+export async function confirmExitWithOrWithoutSaving(captionsToSave: string[], performSave: () => void): Promise<boolean> {
+    let msg: string | HTMLElement = nls.localize('theia/core/quitMessage', 'Any unsaved changes will be lost.');
```

Comment:
I'm not sure I follow the assignment to `| HTMLElement`.

Reply:
Sorry old leftover code. it is meant as a fallback strategy if captionsToSave was left undefined. I'll clean it up a bit.